### ISSUE TITLE
[RFC] Some additional recommendations

### ISSUE THREAD

**Issue by HorlogeSkynet** [Takishima/pylint-secure-coding-standard#5]
Hey @Takishima !

Many thanks for this plugin ; I was about to start working on something similar yesterday, just before I came across yours.

Please find below some additional recommendations/rules I was thinking about :

- [x] Avoid using `os.popen` as well (as [it internally uses `subprocess.Popen` with `shell=True`](https://github.com/python/cpython/blob/3.9/Lib/os.py#L974)) (see #6)
- [x] If the current platform is Windows, avoid using `shlex.quote` (see python/cpython#21502). Of course, the development platform could be different from the production one, but it could be a good start (see #7)
- [x] When using `os.open`, maybe enforce some additional recommendations on `mode` attribute :
    1. ~~Encouraging octal format (?)~~
    2. Discouraging _setuid_, _setgid_ and _sticky bit_
    3. Discouraging any mode offering `W ^ X` (at least for _group_ and _other_)

---

**EDIT**

- [x] #14
- [x] #15
- [x] #9
    - `loop.subprocess_shell` (asyncio)
    - `subprocess.getstatusoutput` & `subprocess.getoutput`

**EDIT 2**

- [x] Avoid using `pickle.load` & `pickle.loads`
- [x] Avoid using `marshal.load` & `marshal.loads`
- [x] Avoid using `shelve.open`

---

Of course, I can contribute (and will if needed), but I wanted first to ask what would you think about that.

Bye 👋

**Comment by Takishima** [comment]
Hi @HorlogeSkynet,

Thanks for your suggestions and comments. 

Since they make sense to me I do not see a reason not to implement those and add them to this project. 

I will try to implement those this week-end, but if you
would like to contribute as well, I would suggest you create a fork and submit a PR. Then we would be able to both commit on the same branch and then finally merge everything once we are happy with the code.

**Comment by HorlogeSkynet** [comment]
- #6 is merge-able _as is_ if you're OK with it
- #7 requires a proper unit test, but I'd rather leave it for you instead of proposing a rag (as I don't know how we can mock `platform.system` with Pytest being called by Pylint's internals)
- about `os.open` permissions, I fear dealing with them statically will be tricky as their internal `mode` computation is based on **current** umask : "[...] When computing *mode*, the current umask value is first masked out"

Anyway, please tweak as you want my branches.
Bye 👋

**Comment by Takishima** [comment]
Regarding `os.open`, I agree that dealing with this statically is most likely not the correct way to go about enforcing this given that it is based on the current umask.

Encouraging the octal format is also going to be troublesome, since the AST only sees a constant integer value (e.g. `420` if `0o644` was passed in).
What we could enforce is if a negative value is passed in as the mode or if the integer value is greater than `0o777 = 511`.

**Comment by HorlogeSkynet** [comment]
> Regarding `os.open`, I agree that dealing with this statically is most likely not the correct way to go about enforcing this given that it is based on the current umask.

👍

> Encouraging the octal format is also going to be troublesome, since the AST only sees a constant integer value (e.g. `420` if `0o644` was passed in).

That's what I feared, thanks for the info.

>  What we could enforce is if a negative value is passed in as the mode or if the integer value is greater than `0o777 = 511`.

Yes, two things from my PoV :

- we can use the octal format in this project code base (more readable in base `8` than in base `10` when it's about UNIX permissions, and integer value will be inferred anyway as you said 😜)
- for `os.open`, regular files are concerned, so we could go for `0o755` to satisfy the `W ^ X` _good practice_  without breaking too many things.

---

**TL; DR** OK for `mode < 0 or mode > 0o755` for me.

---

Bye, thanks again 👋

**Comment by Takishima** [comment]
> > Regarding `os.open`, I agree that dealing with this statically is most likely not the correct way to go about enforcing this given that it is based on the current umask.
> 
> +1
> 
> > Encouraging the octal format is also going to be troublesome, since the AST only sees a constant integer value (e.g. `420` if `0o644` was passed in).
> 
> That's what I feared, thanks for the info.
> 
> > What we could enforce is if a negative value is passed in as the mode or if the integer value is greater than `0o777 = 511`.
> 
> Yes, two things from my PoV :
> 
>     * we can use the octal format in this project code base (more readable in base `8` than in base `10` when it's about UNIX permissions, and integer value will be inferred anyway as you said stuck_out_tongue_winking_eye)
> 
>     * for `os.open`, regular files are concerned, so we could go for `0o755` to satisfy the `W ^ X` _good practice_  without breaking too many things.
> 
> 
> **TL; DR** OK for `mode < 0 or mode > 0o755` for me.
> 
> **EDIT** : [An interesting thread](https://bugs.python.org/issue5220) on `mode` argument. I don't know whether we'll be able to work something out on this subject...
> 
> **EDIT 2** : Actually, I fear _forcing_ users to specify permissions (by discouraging built-in `open`) might be more dangerous than by default, due to this `umask` mess and the fact that dealing with it statically is tricky/impossible. Let me know what you think about this.
> 
> Bye, thanks again wave

I initially developed this plugin to satisfy some of my company's requirements and they want to enforce favoring `os.open` over the builtin `open`... 
That being said, I understand that all of this can be confusing for most users and that in that case it may be better to simply use the builtin function. So what I'll propose is to add an option to this plugin to turn on/off the behavior where we enforce this. Then should the user want to enforce explicitly setting the permissions, we can then simply check `0 < mode < 0o755`

**Comment by HorlogeSkynet** [comment]
> So what I think is best would be to add an option to this plugin to turn on/off the behavior where we enforce this. Then should the user want to enforce explicitly setting the permissions, we then simply check `0 < mode < 0o755`

Very good idea. If I may though : `0 < mode <= 0o755` 😇

**Comment by Takishima** [comment]
Commit [`709b21d`](https://github.com/Takishima/pylint-secure-coding-standard/commit/709b21d99ee3f78f22d945a48bc89f1ab86b7d42) on main branch addresses the builtin `open` vs. `os.open` issue.

Current behaviour is to default to not reporting the `R8009` warning unless the user explicitly sets the `--os-open-mode` configuration value to some truthful value (see `README.md` for more information).
I will add a new warning `W8012` to report the error if `os.open` is used with "unsafe" permissions (based on the `--os-open-mode` config option) later next week.

**Comment by Takishima** [comment]
Commit [09cf3af1](https://github.com/Takishima/pylint-secure-coding-standard/commit/09cf3af1591fbbb25cdf39ca63c3228dd0ca393f) addresses:
- `pickle.load()` and `pickle.loads()`
- `marshal.load()` and `marshal.loads()`
- `shelve.open()`
